src/Serializer/JsonSerializer.php

```diff
@@ -5,48 +5,173 @@
 namespace Patchlevel\EventSourcing\Serializer;
 
 use JsonException;
+use Patchlevel\EventSourcing\Metadata\Event\AttributeEventClassLoader;
+use Patchlevel\EventSourcing\Metadata\Event\AttributeEventMetadataFactory;
+use Patchlevel\EventSourcing\Metadata\Event\EventMetadataFactory;
+use ReflectionClass;
+use TypeError;
 
+use function array_flip;
+use function array_key_exists;
+use function assert;
 use function json_decode;
 use function json_encode;
 
+use const JSON_PRETTY_PRINT;
 use const JSON_THROW_ON_ERROR;
 
 final class JsonSerializer implements Serializer
 {
-    private Hydrator $hydrator;
+    private EventMetadataFactory $metadataFactory;
 
-    public function __construct(?Hydrator $hydrator = null)
+    /** @var array<string, class-string> */
+    private array $eventClassMap;
+
+    /** @var array<class-string, string> */
+    private array $eventClassMapRevert;
+
+    /** @var array<class-string, ReflectionClass> */
+    private array $reflectionClassCache = [];
+
+    /**
+     * @param array<string, class-string> $eventClassMap
+     */
+    public function __construct(EventMetadataFactory $metadataFactory, array $eventClassMap = [])
     {
-        $this->hydrator = $hydrator ?? new DefaultHydrator();
+        $this->metadataFactory = $metadataFactory;
+        $this->eventClassMap = $eventClassMap;
+        $this->eventClassMapRevert = array_flip($eventClassMap);
     }
 
-    public function serialize(object $event): string
+    public function serialize(object $event, array $options = []): SerializedData
```

Comment:
i think we should add here the keys we are using as an doc-type for the options array. So `array{pretty_print: bool}`